src/edu/washington/escience/myria/operator/Sample.java

```diff
@@ -0,0 +1,241 @@
+package edu.washington.escience.myria.operator;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Random;
+import java.util.Set;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+
+import edu.washington.escience.myria.DbException;
+import edu.washington.escience.myria.Schema;
+import edu.washington.escience.myria.Type;
+import edu.washington.escience.myria.storage.TupleBatch;
+import edu.washington.escience.myria.storage.TupleBatchBuffer;
+
+public class Sample extends BinaryOperator {
+  /** Required for Java serialization. */
+  private static final long serialVersionUID = 1L;
+
+  /** Total number of tuples to expect from the right operator. */
+  private int streamSize;
+
+  /** Number of tuples to sample from the right operator. */
+  private int sampleSize;
+
+  /** Random generator used for index selection. */
+  private Random rand;
+
+  /** The type of sampling to perform. Currently supports 'WR' and 'WoR'. */
+  private String sampleType;
+
+  /** True if operator has extracted sampling info. */
+  private boolean computedSamplingInfo = false;
+
+  /** Buffer for tuples that will be returned. */
+  private TupleBatchBuffer ans;
+
+  /** Global count of the tuples seen so far. */
+  private int tupleNum = 0;
+
+  /** Sorted array of tuple indices that will be taken as samples. */
+  private int[] samples;
+
+  /** Current index of the samples array. */
+  private int curSampIdx = 0;
+
+  /**
+   * Instantiate a Sample operator using sampling info from the left operator
+   * and the stream from the right operator.
+   *
+   * @param left
```

Comment:
I would be interested in some background on the design decision to represent the sampling parameters as a single tuple from an arbitrary child operator.